### PR TITLE
cargo check,build,rustc --all-targets

### PR DESCRIPTION
**Pull request by stepancheg**
`cargo check` does not check all targets by default, and to check
all, you need to call it `cargo check --tests --examples --bins
--benches`. `cargo check --all-targets` is a shortcut.

`--all-targets` is added to `check`, `build` and to `rustc` commands.

For consitency `--all-targets` added to other commands like
`test` although "all targets" is default behavior.

### CODE REVIEW

**Comment by alexcrichton** [comment]
cc https://github.com/rust-lang/cargo/issues/2495, https://github.com/rust-lang/cargo/issues/3431

@matklad do you have thoughts about this feature? I'm sort of ambivalent about it.

**Comment by matklad** [comment]
For `cargo check` in particular I think that this should be the default behavior (and if it is the default, we probably don't need a flag). And probably for `build` and `rustc` it is also a good default, but I am not sure that changing default is necessary a good idea at this stage. 

However, just adding a flag as in this PR seems fine to me as well, because the ability to build everything is very useful for development. (But for IntelliJ in particular, we can default our build command to `cargo check --tests --examples --bins --benches` once this flags hit stable).


To sum up: adding flag is fine with me (hope that's the last flag in this area), but what about changing the defaults?

**Comment by malbarbo** [comment]
I think `--all-targets` is misleading because `--target` has a unrelated meaning.

**Comment by stepancheg** [comment]
@malbarbo that's good point, however, these things are already called "targets":

```
    --test NAME                  Check only the specified test target
    --tests                      Check all tests
    --bench NAME                 Check only the specified bench target
    --benches                    Check all benches
```

Any suggestions of better name?

**Comment by behnam** [comment]
This is awesome!

Now that you're on this, maybe it's time to update `Check only the specified [...]` parts in the descriptions (also for other commands) to be more accurate, as these parameters are *union*'ed, meaning each options "adds" to the list of targets, not limiting it to "only" that target. (See also https://github.com/rust-lang/cargo/issues/3112#issuecomment-288890974) IMHO, `Check the specific [...]` is a shorter, more accurate description here, but maybe we can find something even better.

**Comment by alexcrichton** [comment]
@matklad I'd actually argue that we *don't* want this on by default. Especially for `build` and I'd hazard a guess with `check`. I personally have never had the desire for "just build everything" but rather I'm always focused on getting something working. I don't mind adding a flag for this, but it doesn't make a lot of sense to me as a default.

I agree with @malbarbo though that the usage of "target" here can be confusing when mixed with cross-compilation flags. The only other thing I can think of, however, is `--all-artifacts`

**Comment by matklad** [comment]
>I personally have never had the desire for "just build everything" but rather I'm always focused on getting something working.

@alexcrichton sounds reasonable!

Not sure how to deal with ambiguity about targets: it's a fundamental problem in Cargo's terminology, that crate is called target :( 

```
λ cargo build --help | grep target
    --test NAME                  Build only the specified test target
    --bench NAME                 Build only the specified benchmark target
    --target TRIPLE              Build for the target triple
```

I am not sure that adding `artifacts` to the mix will be better then sticking with `target`...

**Comment by Xanewok** [comment]
Recently I encountered the same problem with how to call those crate 'targets'. I assume using "crate type" isn't good, since it's mostly used for output type like `bin`, `lib`, `rlib` etc.? Maybe something along the lines of "crate build type" or "crate target" to disambiguate?

**Comment by alexcrichton** [comment]
I personally think we should stick with `--all-targets` if we do this. It's unfortuante, yes, but there seem to be few other suggestions.

**Comment by stepancheg** [comment]
Another option is to rename it to `--all-build-targets`. And change comment of `--target` to "Build for the _platform_ target triple".

**Comment by alexcrichton** [comment]
That also seems plausible to me, I'm fine either way

**Comment by behnam** [comment]
> Another option is to rename it to --all-build-targets. And change comment of --target to "Build for the platform target triple".

My two cents: in most other compiler/build-systems "build target" almost always means the target platform. Here, it's more of a "target package" or "target crate".

**Comment by matklad** [comment]
@behnam yeah, I think what Cargo calls a "target" is usually known as an "artifact". But we probably can't replace existing Cargo terminology, especially for this single issue, which just exemplifies the general problem.

:+1: to sticking with `--all-targets`. If we ever need all target triples, we could use `--all-target-triples`

**Comment by behnam** [comment]
Yeah, `--all-targets` can definitely work.

At some point, I was thinking that `--all` could be a synonym for `--all-packages`, and what's being added here could also be `--all-package-crates`, `--all-package-targets`, or even `--all-crates`, for short, since from the compiler's perspective, these are called *crates*, and from Cargo perspective, *targets*.

**Comment by alexcrichton** [comment]
@stepancheg would you be willing to rebase this to land?

**Comment by alexcrichton** [comment]
Closing in favor of https://github.com/rust-lang/cargo/pull/4400
